File: LeetCode/463-island-perimeter/ex_463.java
public class Solution {
    public int islandPerimeter(int[][] grid) {
        int count = 0;
        for (int x = 0; x != grid.length; ++x) {
            for (int y = 0; y != grid[0].length; ++y) {
                count += grid[x][y] * 4;
                if (grid[x][y] == 1)
                    count -= adjacent(grid, x, y);
            }
        }
        return count;
    }
    
    private static final int[][] offset = new int[][] { {0, 1}, {1, 0}, {-1, 0}, {0, -1} };

    private int adjacent(int[][] grid, int x, int y) {
        int count = 0;
        for (int i = 0; i < 4; i++) {
            int nx = offset[i][0] + x;
            int ny = offset[i][1] + y;
            if (0 <= nx && nx < grid.length && 0 <= ny && ny < grid[0].length && grid[nx][ny] == 1)
                ++count;
        }
        return count;
    }
}

File: LeetCode/463-island-perimeter/ex_463.py
class Solution(object):
    def islandPerimeter(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        if not grid:    return 0
        count = 0
        for x in range(len(grid)):
            for y in range(len(grid[0])):
                count += grid[x][y] * 4
                if grid[x][y] == 1:
                    count -= self.adjacent(grid, x, y)
        return count
    
    def adjacent(self, grid, x, y):
        offset = [(0, 1), (1, 0), (0, -1), (-1, 0)]
        count = 0
        for i, j in offset:
            nx, ny = x + i, y + j
            if 0 <= nx < len(grid) and 0 <= ny < len(grid[0]) and grid[nx][ny] == 1:
                count += 1
        return count